src/lib/comms/sol-mqtt-impl-mosquitto.c

```diff
@@ -0,0 +1,518 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <mosquitto.h>
+#include <stdbool.h>
+
+#define SOL_LOG_DOMAIN &_sol_mqtt_log_domain
+#include <sol-log-internal.h>
+#include <sol-mainloop.h>
+#include <sol-util.h>
+
+#include "sol-mqtt.h"
+
+SOL_LOG_INTERNAL_DECLARE(_sol_mqtt_log_domain, "mqtt");
+
+static int init_ref;
+#define CHECK_INIT(ret) \
+    do { \
+        if (unlikely(init_ref < 1)) { \
+            SOL_WRN("sol-mqtt used before initialization"); \
+            return ret; \
+        } \
+    } while (0)
+
+#define MQTT_CHECK_API(ptr, ...) \
+    do {                                        \
+        if (unlikely(ptr->api_version != \
+            SOL_MQTT_CONFIG_API_VERSION)) { \
+            SOL_WRN("Couldn't handle mqtt handler that has unsupported " \
+                "version '%u', expected version is '%u'", \
+                ptr->api_version, SOL_MQTT_CONFIG_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0)
+
+struct sol_mqtt {
+    struct mosquitto *mosq;
+
+    struct sol_fd *socket_watch;
+
+    struct sol_timeout *connect_timeout;
+    struct sol_timeout *disconnect_timeout;
+    struct sol_timeout *publish_timeout;
+    struct sol_timeout *subscribe_timeout;
+    struct sol_timeout *unsubscribe_timeout;
+    struct sol_timeout *message_timeout;
+
+    void *data;
+
+    struct sol_mqtt_handlers handlers;
+
+    int socket_fd;
+
+    int connection_status;
+
+    time_t keepalive;
+};
+
+static struct sol_mqtt_message *
+sol_mqtt_message_new(const char *topic, const struct sol_buffer *payload, sol_mqtt_qos qos, bool retain)
+{
+    struct sol_mqtt_message *message;
+
+    SOL_NULL_CHECK(topic, NULL);
+    SOL_NULL_CHECK(payload, NULL);
+
+    message = calloc(1, sizeof(struct sol_mqtt_message));
+    SOL_NULL_CHECK(message, NULL);
+
+    message->topic = sol_util_memdup(topic, strlen(topic) + 1);
+    SOL_NULL_CHECK_GOTO(message->topic, topic_error);
+
+    message->payload = sol_buffer_copy(payload);
+    SOL_NULL_CHECK_GOTO(message->payload, payload_error);
+
+    message->qos = qos;
+    message->retain = retain;
+
+    return message;
+
+payload_error:
+    free(message->topic);
+topic_error:
+    free(message);
+
+    return NULL;
+}
+
+static void
+sol_mqtt_message_free(struct sol_mqtt_message *message) {
+    if (!message)
+        return;
+
+    sol_buffer_fini(message->payload);
+    free(message->payload);
+    free(message->topic);
+    free(message);
+}
+
+static void
+sol_mqtt_init(void)
+{
+    SOL_LOG_INTERNAL_INIT_ONCE;
+    init_ref++;
+    mosquitto_lib_init();
```

Comment:
Ah yes, my bad. It shouldn't be a problem but there is no need to call it again anyway. I'll fix it. Thanks.